Unable to decode error JSON when sending exception details over proxy

**Description**
When uploading data object if disk out of space, it throws a json decoding error.
This usually occurs when the expected JSON is a normal string e.g. "this is an error", missing the required "{" character.

```python
2023-07-19 02:14:03.589 [ERROR] [analysis:d77f54be60df2984d388bc55197952bb798c1f3d72f6a5129ecb6521a332fce1:district-cooling] failed -> Traceback (most recent call last):
  File "/mnt/storage/dev/venv-servers/lib/python3.10/site-packages/requests/models.py", line 971, in json
    return complexjson.loads(self.text, **kwargs)
  File "/mnt/storage/dev/venv-servers/lib/python3.10/site-packages/simplejson/__init__.py", line 514, in loads
    return _default_decoder.decode(s)
  File "/mnt/storage/dev/venv-servers/lib/python3.10/site-packages/simplejson/decoder.py", line 386, in decode
    obj, end = self.raw_decode(s)
  File "/mnt/storage/dev/venv-servers/lib/python3.10/site-packages/simplejson/decoder.py", line 416, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/mnt/storage/dev/venv-servers/lib/python3.10/site-packages/saas/rest/proxy.py", line 31, in extract_response
    content = response.json()
  File "/mnt/storage/dev/venv-servers/lib/python3.10/site-packages/requests/models.py", line 975, in json
    raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/mnt/storage/dev/venv-servers/lib/python3.10/site-packages/explorer/analysis/base.py", line 276, in run
    results = self._analysis.perform_analysis(self._group, self._scene, self)
  File "/mnt/storage/dev/venv-servers/lib/python3.10/site-packages/duct/analyses/district_cooling.py", line 76, in perform_analysis
    bf_obj = context.sdk.upload_content(building_footprints_path, 'DUCT.GeoVectorData', 'geojson', False)
  File "/mnt/storage/dev/venv-servers/lib/python3.10/site-packages/saas/sdk/base.py", line 381, in upload_content
    meta = dor.add_data_object(content_path, self._authority.identity, access_restricted, content_encrypted,
  File "/mnt/storage/dev/venv-servers/lib/python3.10/site-packages/saas/dor/proxy.py", line 72, in add_data_object
    result = self.post('add-c', body=body, attachment_path=content_path)
  File "/mnt/storage/dev/venv-servers/lib/python3.10/site-packages/saas/rest/proxy.py", line 246, in post
    return extract_response(response)
  File "/mnt/storage/dev/venv-servers/lib/python3.10/site-packages/saas/rest/proxy.py", line 37, in extract_response
    raise SaaSRuntimeException(response.reason, details={
saas.core.exceptions.SaaSRuntimeException: Internal Server Error
```

--- a/saas/core/exceptions.py
+++ b/saas/core/exceptions.py
@@ -1,8 +1,12 @@
+import json
 from typing import Optional
 
 from pydantic import BaseModel, Field
 
 from saas.core.helpers import generate_random_string
+from saas.core.logging import Logging
+
+logger = Logging.get('saas.core')
 
 
 class ExceptionContent(BaseModel):
@@ -17,6 +21,20 @@ class ExceptionContent(BaseModel):
 class SaaSRuntimeException(Exception):
     def __init__(self, reason: str, details: dict = None, id: str = None):
         self._content = ExceptionContent(id=id if id else generate_random_string(16), reason=reason, details=details)
+
+        # check if the details can be JSON encoded
+        try:
+            json.dumps(self._content.dict())
+        except TypeError:
+            logger.warning(f"Encountered JSON incompatible exception details: class={self.__class__.__name__} "
+                           f"id={self._content.id} details: {self._content.details}")
+
+            # convert detail values into strings as a fallback
+            self._content.details = {
+                k: str(v) for k, v in self._content.details.items()
+            }
+        except Exception as e:
+            logger.error(f"Encountered JSON incompatible exception content: {e} {reason} {details}")
 
     @property
     def id(self):

--- a/tests/test_core.py
+++ b/tests/test_core.py
@@ -3,6 +3,7 @@ import os
 
 import pytest
 
+from saas.core.exceptions import SaaSRuntimeException
 from saas.core.schemas import GithubCredentials, SSHCredentials
 from saas.core.keystore import Keystore
 from saas.core.eckeypair import ECKeyPair
@@ -314,3 +315,25 @@ def test_rollover(logging, temp_directory):
     assert(os.path.isfile(log_path0))
     assert(os.path.isfile(log_path1))
     assert(os.path.isfile(log_path2))
+
+
+def test_json_incompatible_exception():
+    class SomeClass:
+        pass
+
+    instance = SomeClass()
+
+    # create an exception with details that cannot be JSON encoded. it should still work but everything in details
+    # is turned into a string.
+    e = SaaSRuntimeException('something happened...', details={
+        'a': 34,
+        'b': 4.5,
+        'c': 'sdfs',
+        'd': {
+            'd1': 'sd'
+        },
+        'e': instance
+    })
+
+    for k, v in e.details.items():
+        assert isinstance(v, str)

